### PR TITLE
Selecting better center for higher chances of booking in one pass

### PR DESCRIPTION
**Pull request by BlackRider97**
1. Trying out all centers instead of selecting one again checking status of centers
2. Random slots of a center for handling concurrency cases of same script run for a same region

### CODE REVIEW

**Comment by bombardier-gif** [comment]
This is a really cool idea tbh. Can we test it a bit more and then I will merge this. 
There have been lot of bugs in changes in the core flow.

**Comment by bombardier-gif** [comment]
@BumpyKnuckls FYI

**Comment by BumpyKnuckls** [comment]
Thanks for the tag @bombardier-gif Will take a look.

**Comment by BlackRider97** [comment]
Yes I'll try this with at-least two bookings.  One pass is succeeded so far but that may be a luck

**Comment by bombardier-gif** [comment]
Thanks let me know, and then i will merge it.

**Comment by BlackRider97** [comment]
Also fixed Internal Server error case at time of high booking. This will help irrespective of booking strategy.

**Comment by bombardier-gif** [comment]
@BlackRider97 what does it do when it gets the server error?

**Comment by shailesh** [comment]
Screenshot 2021-05-21 at 8 35 43 AM

Please test and fix this.

**Comment by BlackRider97** [comment]
>  what does it do when it gets the server error?

@bombardier-gif  It retries with same center & same slot again

**Comment by BlackRider97** [comment]
@shailesh I guess you are looking at older commits. I have fixed it yesterday only

**Comment by bombardier-gif** [comment]
Cool. There is some issue in https://github.com/bombardier-gif/covid-vaccine-booking/issues/240
As soon as its fixed, i will merge this.

**Comment by shailesh** [comment]
> Cool. There is some issue in #240
> As soon as its fixed, i will merge this.

#240 was about the latest merge not related to this one. I'll test this later to-night

**Comment by bombardier-gif** [comment]
yes i will wait till critical bugs are fixed... the more PRs we merge, the more issues we get in first 2-3 days

**Comment by shailesh** [comment]
@BlackRider97 have you tested this?

**Comment by BlackRider97** [comment]
No slots were open in my region. Still waiting to test one more time

File: src/utils.py [hunk 1]
```diff
@@ -478,13 +481,28 @@ def book_appointment(request_header, details, mobile, generate_captcha_pref, api
                 os.system("pause")
                 sys.exit()
 
+            elif resp.status_code == 409:
+                print(f"Response: {resp.status_code} : {resp.text}")
+                try:
+                    data = resp.json()
+                    # Response: 409 : {"errorCode":"APPOIN0040","error":"This vaccination center is completely booked for the selected date. Please try another date or vaccination center."}
+                    if data.get("errorCode", '') == 'APPOIN0040':
```
Comment (shailesh) [inline src/utils.py:489]: make sure about checks on these error codes and confirm if these are not dynamic ones.

**Review by shailesh — COMMENTED** [review]
> No slots were open in my region. Still waiting to test one more time

check for Jharkhand ---> Garhwa

**Comment by shailesh** [comment]
At least happy scenarios should work for non-happy pick up any state and district like Delhi, Hyderabad, Bangalore,

File: src/utils.py (hunk 1 above)
Comment (BlackRider97) [reply src/utils.py:489]: Yes error codes are fixed ony

**Comment by shailesh** [comment]
Works fine for the happy cases. @bombardier-gif

**Review by shailesh — APPROVED** [review]

**Comment by bombardier-gif** [comment]
Great! Can you please resolve the conflicts, I will go ahead and merge

**Comment by shailesh** [comment]
@BlackRider97 please resolve the conflicts, there must be just white spaces thing. commit and then we can merge this. I have been using this for a couple of days now. Works fine

**Comment by shailesh** [comment]
@Nakul93 can you resolve this too.

**Comment by Nakul93** [comment]
> @Nakul93 can you resolve this too.

I don't have access

**Comment by shailesh** [comment]
> > @Nakul93 can you resolve this too.
> 
> I don't have access

got it

**Comment by BlackRider97** [comment]
Resolved conflicts. It works in happy and case where it is performing better than current logic

File: src/utils.py [hunk 2]
```diff
@@ -558,56 +577,106 @@ def check_and_book(
...
+                            "beneficiaries": [
+                                beneficiary["bref_id"] for beneficiary in beneficiary_dtls
+                            ],
+                            "dose": dose_num,
+                            "center_id": option["center_id"],
+                            "session_id": option["session_id"],
+                            "slot": selected_slot,
+                        }
+                        print(f"Booking with info: {new_req}")
+                        booking_status = book_appointment(request_header, new_req, mobile, captcha_automation, captcha_automation_api_key, captcha_api_choice)
```
Comment (shailesh) [inline src/utils.py:663]: this should not have gone

Comment (shailesh) [reply src/utils.py:663]: remove this one will resolve #259

**Review by shailesh — COMMENTED** [review]
`booking_status = book_appointment(request_header, new_req, mobile, captcha_automation)`

**Pull request merged**